### PR TITLE
ReNA: fast agglomerative clustering

### PR DESCRIPTION
**Pull request by ahoyosid**
Fast agglomerative clustering based on recursive nearest neighbor grouping.
ReNA yields very fast and accurate models, enabling to process large datasets on budget.
Unlike existing fast agglomerative schemes, it avoids the creation of giant clusters.

parcellation
original
compression

Note: I put this in connectivity just to be consistent with the Ward clustering example, I'll probably put it in a "cluster" folder. 
Any ideas ?

### CODE REVIEW

**Comment by ahoyosid** [comment]
Can someone check this PR out ?
@KamalakerDadi, @mrahim, @lesteve

File: examples/03_connectivity/plot_fast_clustering.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+The big picture discussion corresponding to this example can be found
+in the documentation section :ref:`parcellating_brain`.
+"""
+
+##################################################################
+# Download a rest dataset and turn it to a data matrix
+# -----------------------------------------------------
+#
+# We we download one subject of the ADHD dataset from Internet
```
Comment (AlexandreAbraham) [inline examples/03_connectivity/plot_fast_clustering.py]: No wewe in the code, please!

File: examples/03_connectivity/plot_fast_clustering.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#
+# First we display the labels of the clustering in the brain.
+#
+# To visualize results, we need to transform the clustering's labels back
+# to a neuroimaging volume. For this, we use the NiftiMasker's
+# inverse_transform method.
+from nilearn.plotting import plot_roi, plot_epi, show
+import numpy as np
+
+# Unmask the labels
```
Comment (AlexandreAbraham) [inline examples/03_connectivity/plot_fast_clustering.py]: Misplaced comment ?

File: nilearn/connectome/rena_clustering.py [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+        Weights corresponding to all edges in the mask.
+        shape: (n_edges,)
+    """
+    data = masker.inverse_transform(masked_data).get_data()
+
+    weights_deep = np.sum(np.diff(data, axis=2) ** 2, axis=-1).ravel()
+    weights_right = np.sum(np.diff(data, axis=1) ** 2, axis=-1).ravel()
+    weights_down = np.sum(np.diff(data, axis=0) ** 2, axis=-1).ravel()
+
+    weights = np.hstack([weights_deep, weights_right, weights_down])
```
Comment (AlexandreAbraham) [inline nilearn/connectome/rena_clustering.py]: Nitpicking: do you think that we should preallocate this?

File: examples/03_connectivity/plot_fast_clustering.py (hunk 1 above)
Comment (ahoyosid) [inline examples/03_connectivity/plot_fast_clustering.py]: ok

File: examples/03_connectivity/plot_fast_clustering.py (hunk 2 above)
Comment (ahoyosid) [inline examples/03_connectivity/plot_fast_clustering.py]: indeed

File: nilearn/connectome/rena_clustering.py (hunk 3 above)
Comment (ahoyosid) [inline nilearn/connectome/rena_clustering.py]: I'm not sure, we build the weights once

File: nilearn/connectome/rena_clustering.py [hunk 4]
```diff
@@ -0,0 +1,638 @@
...
+
+    return edges, weights, edges_mask
+
+
+def weighted_connectivity_graph(masker, masked_data):
+    """ Creating weighted graph: data and topology are encoded by a
+    connectivity matrix.
+
+    Parameters
+    ----------
```
Comment (mrahim) [inline nilearn/connectome/rena_clustering.py]: Input `masker` description is missing

File: nilearn/connectome/rena_clustering.py [hunk 5]
```diff
@@ -0,0 +1,638 @@
...
+        In practice it is 1e-7.
+
+    Returns
+    -------
+    reduced_connectivity : a sparse matrix in COOrdinate format.
+
+    reduced_masked_data: array like
+        2D data matrix.
+    """
+    n_features = len(labels)
```
Comment (mrahim) [inline nilearn/connectome/rena_clustering.py]: This is a bit confusing since you have also `n_labels` as input

File: nilearn/connectome/rena_clustering.py [hunk 6]
```diff
@@ -0,0 +1,638 @@
...
+
+def _reduce_data_and_connectivity(labels, n_labels, connectivity, masked_data,
+                                  threshold):
+    """Perform feature grouping and reduce the connectivity matrix.
+
+    Parameters
+    ----------
+    labels : array like
+
+    n_labels : int
```
Comment (mrahim) [inline nilearn/connectome/rena_clustering.py]: Could you explain more ?

File: nilearn/connectome/rena_clustering.py [hunk 7]
```diff
@@ -0,0 +1,638 @@
...
+
+
+def recursive_nearest_agglomeration(masker, masked_data, n_clusters, n_iter,
+                                    threshold):
+    """Recursive nearest agglomeration: it performs iteratively the nearest
+    neighbor grouping.
+
+    Parameters
+    ----------
+    connectivity : a sparse matrix in COOrdinate format.
```
Comment (mrahim) [inline nilearn/connectome/rena_clustering.py]: There is `masker` parameter instead of `connectivity`

File: nilearn/connectome/rena_clustering.py [hunk 8]
```diff
@@ -0,0 +1,638 @@
...
+        labels = reduced_labels[labels]
+        n_labels = connectivity.shape[0]
+
+        if n_labels <= n_clusters:
+            break
+
+    return n_labels, labels
+
+
+class ReNA(BaseEstimator):
```
Comment (mrahim) [inline nilearn/connectome/rena_clustering.py]: Inherit also from `TransformerMixin` and `ClusterMixin`

File: nilearn/connectome/rena_clustering.py [hunk 9]
```diff
@@ -0,0 +1,638 @@
...
+        _, inverse = np.unique(self.labels_, return_inverse=True)
+
+        if self.scaling:
+            Xred = Xred / np.sqrt(self.sizes_)
+        X_inv = Xred[..., inverse]
+
+        return self.masker_.inverse_transform(X_inv)
+
+
+# XXX this code is also replicated in the Metaestimator PR
```
Comment (mrahim) [inline nilearn/connectome/rena_clustering.py]: `XXX` to be addressed ?

File: examples/03_connectivity/plot_fast_clustering.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+fmri_masked = nifti_masker.transform(func_filename)
+
+# Display the original data
+plot_epi(nifti_masker.inverse_transform(fmri_masked[0]),
+         title='Original (%i voxels)' % fmri_masked.shape[1],
+         cut_coords=cut_coords, vmax=fmri_masked.max(), vmin=fmri_masked.min(),
+         display_mode='xz')
+
+
+# A reduced data can be create by taking the parcel-level average:
```
Comment (bthirion) [inline examples/03_connectivity/plot_fast_clustering.py]: 'created'

File: examples/03_connectivity/plot_fast_clustering.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+# Note that, as many objects in the scikit-learn, the ReNA object exposes
```
Comment (bthirion) [inline examples/03_connectivity/plot_fast_clustering.py]: 'as many scikit-learn objects',

File: nilearn/connectome/rena_clustering.py [hunk 12]
```diff
@@ -0,0 +1,626 @@
...
+    weights_deep = np.sum(np.diff(data, axis=2) ** 2, axis=-1).ravel()
+    weights_right = np.sum(np.diff(data, axis=1) ** 2, axis=-1).ravel()
+    weights_down = np.sum(np.diff(data, axis=0) ** 2, axis=-1).ravel()
+
+    weights = np.hstack([weights_deep, weights_right, weights_down])
+
+    return weights
+
+
+def _make_edges_3d(vertices, is_mask):
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: _make_3d_edges

File: nilearn/connectome/rena_clustering.py [hunk 13]
```diff
@@ -0,0 +1,626 @@
...
+    masker : NiftiMasker
+        The nifti masker used to mask the data.
+
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix.
+
+    Returns
+    -------
+    edges : numpy array
+
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: Please explain a  little bit...

File: nilearn/connectome/rena_clustering.py [hunk 14]
```diff
@@ -0,0 +1,626 @@
...
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix
+
+    Returns
+    -------
+    connectivity : a sparse COO matrix
+    """
+    n_features = masker.mask_img_.get_data().sum()
+
+    edges, weight, edges_mask = _make_edges_and_weights(masker, masked_data)
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: edges_mask is never used ?

File: nilearn/connectome/rena_clustering.py [hunk 15]
```diff
@@ -0,0 +1,626 @@
...
+
+    if n_labels < n_clusters:
+        # cut some links to achieve the desired number of clusters
+        alpha = n_features - n_clusters
+
+        nn_connectivity = nn_connectivity + nn_connectivity.T
+
+        edges_ = np.array(nn_connectivity.nonzero())
+
+        plop = edges_[0] - edges_[1]
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: I am probably guilty for this unfortunate naming, but this should certainly be changed...

File: nilearn/connectome/rena_clustering.py [hunk 16]
```diff
@@ -0,0 +1,626 @@
...
+
+        select = np.argsort(plop)[:alpha]
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: Explain a little bit what the algorithm is doing here: remove edges so that the final number of clusters is not less than n_clusters.

File: nilearn/connectome/rena_clustering.py [hunk 17]
```diff
@@ -0,0 +1,626 @@
...
+    ----------
+    masker : NiftiMasker instance
+
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix
+
+    n_clusters : int
+        The number of clusters to find.
+
+    n_iter : int
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: Could get a reasonable default  value ?

File: nilearn/connectome/rena_clustering.py [hunk 18]
```diff
@@ -0,0 +1,626 @@
...
+        Number of iterations.
+
+    threshold : float in the close interval [0, 1]
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: Could get a reasonable default value ?

File: nilearn/connectome/rena_clustering.py [hunk 19]
```diff
@@ -0,0 +1,626 @@
...
+
+        if n_labels <= n_clusters:
+            break
+
+    return n_labels, labels
+
+
+class ReNA(BaseEstimator, ClusterMixin, TransformerMixin):
+    """
+    Recursive nearest agglomeration.
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: "Recursive Neighbor Agglomeration"

File: nilearn/connectome/rena_clustering.py [hunk 20]
```diff
@@ -0,0 +1,626 @@
...
+        documentation for details
+
+    t_r : float, optional (default None)
+        This parameter is passed to signal.clean. Please see the related
+        documentation for details.
+
+    n_iter : int, optional (default 10)
+        Number of iterations of the recursive nearest agglomeration
+
+    threshold : float in the opened interval (0., 1.), optional (default 1e-7)
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: open interval

File: nilearn/connectome/rena_clustering.py [hunk 21]
```diff
@@ -0,0 +1,626 @@
...
+        """
+
+        check_is_fitted(self, "masker_")
+        check_is_fitted(self, "labels_")
+
+        X = self.masker_.transform(X)
+
+        unique_labels = np.unique(self.labels_)
+
+        nX = []
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: better naming ?

File: nilearn/connectome/rena_clustering.py [hunk 22]
```diff
@@ -0,0 +1,626 @@
...
+        for l in unique_labels:
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: for label in unique_labels:

File: nilearn/connectome/rena_clustering.py [hunk 23]
```diff
@@ -0,0 +1,626 @@
...
+    def inverse_transform(self, Xred):
+        """
+        Parameters
+        ----------
+        Xred : numpy array
+            2D data matrix of shape [n_samples, n_features]
+
+        Returns
+        -------
+        X_inv : Niimg
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: Please be a bit more descriptive in the official API...

File: nilearn/connectome/rena_clustering.py [hunk 24]
```diff
@@ -0,0 +1,626 @@
...
+
+        if self.scaling:
+            Xred = Xred / np.sqrt(self.sizes_)
+        X_inv = Xred[..., inverse]
+
+        return self.masker_.inverse_transform(X_inv)
+
+
+# XXX this code is also replicated in the Metaestimator PR
+def _check_masking(mask, smoothing_fwhm, target_affine, target_shape,
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: Should probably be somewhere else ?

File: nilearn/connectome/tests/test_rena.py [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+
+    assert_equal(10, rena.n_clusters_)
+
+    memory = Memory(cachedir=None)
+    rena2 = ReNA(n_clusters=-2, mask=nifti_masker, memory=memory)
+    assert_raises(ValueError, rena2.fit, data)
+
+    rena3 = ReNA(n_clusters=10, mask=None, scaling=True)
+    data_red2 = rena3.fit_transform(index_img(data, 0))
+    data_compress2 = rena3.inverse_transform(data_red2)
```
Comment (bthirion) [inline nilearn/connectome/tests/test_rena.py]: Check a few other parameters, e.g. n_iter. 
Check that the returned values are fine etc.

**Comment by ahoyosid** [comment]
@bthirion  can you check this PR out again?

File: nilearn/connectome/rena_clustering.py [hunk 26]
```diff
@@ -0,0 +1,646 @@
...
+
+    masked_data : numpy array of shape [n_samples, n_features]
+        Image in brain space transformed into 2D data matrix
+
+    n_clusters : int
+        The number of clusters to find.
+
+    threshold : float in the close interval [0, 1]
+        The treshold is setted to handle eccentricities.
+        In practice it is 1e-7.
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: again

File: nilearn/connectome/rena_clustering.py [hunk 27]
```diff
@@ -0,0 +1,646 @@
...
+
+    n_labels = n_features - (nn_connectivity + nn_connectivity.T).nnz / 2
+
+    if n_labels < n_clusters:
+        # remove edges so that the final number of clusters is not less than
+        # n_clusters (to achieve the desired number of clusters)
+        n_edges = n_features - n_clusters
+
+        nn_connectivity = (nn_connectivity + nn_connectivity.T)
+
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: Maybe that's a matter of taste, but i don't like skipping lines without reason. I do it after an indented block or whenever pep8 requires, but not here.

File: nilearn/connectome/rena_clustering.py [hunk 28]
```diff
@@ -0,0 +1,646 @@
...
+class ReNA(BaseEstimator, ClusterMixin, TransformerMixin):
+    """Recursive Neighbor Agglomeration:
+    Recursively merges the pair of clusters according to 1-nearest neighbors
+    criterion.
+
+    Parameters
+    ----------
+    mask : filename, niimg, NiftiMasker instance, optional default None)
+        Mask to be used on data. If an instance of masker is passed,
+        then its mask will be used. If no mask is it will be computed
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: If no mask is passed,

File: nilearn/connectome/rena_clustering.py [hunk 29]
```diff
@@ -0,0 +1,646 @@
...
+
+    `labels_ ` : array-like, (n_features,)
+        cluster labels for each feature.
+
+    `n_clusters_` : int
+        Number of clusters.
+
+    `sizes_` : array-like (n_features,)
+        It contains the size of each cluster.
+    """
```
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: A suggestion here: in the builder signature, group together the parameters that are specific to ReNA on one hand, and those that are common to all maskers on the other hand.

File: nilearn/connectome/rena_clustering.py (hunk 24 above)
Comment (bthirion) [inline nilearn/connectome/rena_clustering.py]: ?

**Comment by kchawla-pi** [comment]
Is this alive? If not, please close it by this weekend.
